### PR TITLE
feat: Bootstrap the mini-controller

### LINKED ISSUE #588: Bootstrap the mini-controller for the branch-brased planner
Acceptance criteria:
* There is a new `planOnly bool` field on the `Terraform` object. ([see comments here][disc-1])
* There is a controller that does nothing.
* There is a single binary that runs the mini controller and the webhook receiver in separate go routines.

[disc-1]: https://github.com/weaveworks/weave-gitops-enterprise/issues/2809

### PR DESCRIPTION
**Pull request by yitsushi**
* There is a new field on Terraform: planOnly. It is not used anywhere,
  yet, tf controller ignores it.
* `webhook-receiver` is called `branch-based-planner` now, as the binary
  is not only the webhook receiver, this name fits more the
  functionality.
* The informer does nothing yet, it's just a bootstrap phase.
* The Informer and the Webhook receiver run in their own go routine, if
  any of them stops, it terminates the other as none of them are really
  useful without their parner.

Closes #588

References:
* https://github.com/weaveworks/tf-controller/issues/588

### CODE REVIEW

**Comment by chanwit** [comment]
> The mini controller does nothing yet, it's just a bootstrap phase.

To simplify thing, we don't need any stuff like the reconciliation loop in the mini controller. 

The mini controller would be more like:
  - identifying target objects to be cloned, cloning set of objects per spec. the information would be obtained from the PR.
  - setting up an informer to watch Terraform CR objects with `planOnly=true` (you would also put labels during the object cloning procedure to make them query-able).

**Comment by yitsushi** [comment]
> To simplify thing, we don't need any stuff like the reconciliation loop in the mini controller.

Something has to watch the resources to fetch the plan output and post it as a comment.

**Comment by chanwit** [comment]
> Something has to watch the resources to fetch the plan output and post it as a comment, not?

exactly: so it's simpler to implement the controller like this with an informer:

> setting up an informer to watch Terraform CR objects with planOnly=true (you would also put labels during the object cloning procedure to make them query-able).

**Comment by yitsushi** [comment]
Ah makes sense. Thank you for clarifying, my eyes skipped the "informer" part and recognised only the "don't need any stuff like the reconciliation loop".

File: cmd/branch-based-planner/main.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			log.Error(err, "unable to start webhook server")
+		}
+
+		// Does not matter if it was an error or not, if this routine is done for
+		// unknown reasons, stop the other routine too.
+		informerCancel()
+	}(log.WithName("webhook-server"))
+
+	func(log logr.Logger) {
+		log.Info("Starting branch-blased planner informer")
```
Comment (yiannistri) [inline cmd/branch-based-planner/main.go:81]:
```suggestion
		log.Info("Starting branch-based planner informer")
```

File: cmd/branch-based-planner/main.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	go func(log logr.Logger) {
+		log.Info("Starting webhook server")
+
+		if err := startWebhookServer(webhookCtx, log); err != nil {
+			log.Error(err, "unable to start webhook server")
+		}
+
+		// Does not matter if it was an error or not, if this routine is done for
+		// unknown reasons, stop the other routine too.
+		informerCancel()
```
Comment (yiannistri) [inline cmd/branch-based-planner/main.go:77]:
```suggestion
		webhookCancel()
```

File: cmd/branch-based-planner/main.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	func(log logr.Logger) {
+		log.Info("Starting branch-blased planner informer")
+
+		if err := startInformer(informerCtx, log, clusterClient); err != nil {
+			log.Error(err, "unable to start branch-based planner informer")
+		}
+
+		// Does not matter if it was an error or not, if this routine is done for
+		// unknown reasons, stop the other routine too.
+		webhookCancel()
```
Comment (yiannistri) [inline cmd/branch-based-planner/main.go:89]:
```suggestion
		informerCancel()
```

File: controllers/utils.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+	tfv1alpha2 "github.com/weaveworks/tf-controller/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func IndexBy(kind string) func(o client.Object) []string {
+	return func(o client.Object) []string {
+		terraform, ok := o.(*tfv1alpha2.Terraform)
+		if !ok {
+			panic(fmt.Sprintf("Expected a Kustomization, got %T", o))
```
Comment (yiannistri) [inline controllers/utils.go:14]:
```suggestion
			panic(fmt.Sprintf("Expected a Terraform object, got %T", o))
```

File: controllers/tf_bbp_controller.go
Comment (yiannistri) [inline controllers/tf_bbp_controller.go:1]: Small typo `TerraformBranchBasesPlannerReconciler` -> `TerraformBranchBasedPlannerReconciler`

**Review by yiannistri — COMMENTED** [review]
Some minor typos but generally looks fine 👍

File: cmd/branch-based-planner/main.go (hunk 3 above)
Comment (yitsushi) [reply cmd/branch-based-planner/main.go:89]: That's intentional. If the informer stops, cancel webhook context, if the webhook stops, cancel informer context.

File: cmd/branch-based-planner/main.go (hunk 2 above)
Comment (yitsushi) [reply cmd/branch-based-planner/main.go:77]: That's intentional. If the informer stops, cancel webhook context, if the webhook stops, cancel informer context.

File: controllers/tf_bbp_controller.go
Comment (yitsushi) [reply controllers/tf_bbp_controller.go:1]: Actually that whole block can be deleted as we don't have a controller anymore.

File: cmd/branch-based-planner/main.go (hunk 2 above)
Comment (yiannistri) [reply cmd/branch-based-planner/main.go:77]: Ah I didn't realise.

**Review by yiannistri — DISMISSED** [review]
Looks good 👍

File: controllers/utils.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+			namespace := terraform.GetNamespace()
+			if terraform.Spec.SourceRef.Namespace != "" {
+				namespace = terraform.Spec.SourceRef.Namespace
+			}
+			return []string{fmt.Sprintf("%s/%s", namespace, terraform.Spec.SourceRef.Name)}
+		}
+
+		return nil
+	}
+}
```
Comment (chanwit) [inline controllers/utils.go:27]: Please revert this too.

**Review by chanwit — CHANGES_REQUESTED** [review]

File: cmd/branch-based-planner/main.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	}
+
+	if err := server.Start(ctx); err != nil {
+		return fmt.Errorf("problem running the webhook receiver server: %w", err)
+	}
+
+	return nil
+}
+
+func setupCerts(opts *applicationOptions) (chan struct{}, *mtls.CertRotator) {
```
[the remark this answers is not in the document]
Comment (yitsushi) [reply cmd/branch-based-planner/main.go]: Yes it's a leftover from the controller.

File: controllers/utils.go (hunk 5 above)
Comment (yitsushi) [reply controllers/utils.go]: done

File: controllers/tf_controller.go [hunk 7]
```diff
@@ -425,19 +425,19 @@ func shortRev(revision string) string {
...
 	if err := mgr.GetCache().IndexField(context.TODO(), &infrav1.Terraform{}, infrav1.BucketIndexKey,
-		r.IndexBy(sourcev1b2.BucketKind)); err != nil {
+		IndexBy(sourcev1b2.BucketKind)); err != nil {
 		return fmt.Errorf("failed setting index fields: %w", err)
 	}
 
 	// Index the Terraforms by the OCIRepository references they (may) point at.
 	if err := mgr.GetCache().IndexField(context.TODO(), &infrav1.Terraform{}, infrav1.OCIRepositoryIndexKey,
-		r.IndexBy(sourcev1b2.OCIRepositoryKind)); err != nil {
+		IndexBy(sourcev1b2.OCIRepositoryKind)); err != nil {
```
[the remark this answers is not in the document]
Comment (yitsushi) [reply controllers/tf_controller.go]: done

**Comment by yitsushi** [comment]
@chanwit : removed unused (leftover from the Controller logic) code blocks and reverted the `IndexBy` function to be a function on the `TerraformReconciler` struct again, but fixed the error message @yiannistri spotted, so that's still a change in the `tf_controller.go` file.

**Review by chanwit — APPROVED** [review]
LGTM, thank you @yitsushi :1st_place_medal:

**Pull request merged**